content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
++++
+title = "An Efficient Implementation of Self"
+
+[extra]
+bio = """
+  [Rachit Nigam](https://rachitnigam.com) is a second year PhD student interested in
+  programming languages & computer architecture. In his free time, he
+  [subtweets](https://twitter.com/notypes/status/1170037148290080771) his advisor and [annoys tenured professors](https://twitter.com/natefoster/status/1074401015565291520).
+"""
+[[extra.authors]]
+name = "Rachit Nigam"
+link = "https://rachitnigam.com"
++++
+
+[An Efficient Implementation of Self, a Dynamically-Typed Object-Oriented
+Language Based on Prototypes][paper] presents techniques for runtime
+compilation, now more commonly referred to as Just In Time (JIT) compilation,
+for the [Self][] programming language.
+
+> We have developed and implemented techniques that double the performance of
+> the dynamically-typed object oriented languages.
+
+
+## The Challenge
+
+The term "dynamic language" is mostly commonly associated with modern scripting
+languages like Python and JavaScript. Self, a much older language developed
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
+extreme -- *everything* in self is a message to an object. This includes
+Java-like method calls on object *as well as* control structures like loops
+and conditionals.
+
+For example, the Self condition `IfTrue:IfFalse:` is a method invocation on
+the boolean object `true` and `false`. This means that a simple conditional:
+
+```
+if (x) 1 else 2
+```
+
+in Self parlance (written with JavaScript-esque syntax) looks like this:
```

Comment:
The above is a little hard to understand for two reasons:

- It's not clear which of these is supposed to be JavaScript-like. I think it's the second one. But then it's not 100% sure what "language" the first snippet is written in—I think it's just a notional, abstract language, but maybe it's worth mentioning that it's not supposed to be Self or Self-esque JS.
- In JS, at least `if (x) 1 else 2` isn't terribly meaningful because `if` isn't an expression, unlike `x ? 1 : 2`. Maybe that's fine, though—this does make sense in, like, Scala or Rust.